### PR TITLE
Add Leader Election module to noobaa core

### PR DESCRIPTION
**Pull request by nadavMiz**
### Describe the Problem
Today the NooBaa core pod runs as a single active instance. If it fails or is rescheduled, there is no built-in way for a standby core pod to take over cleanly. We need active-passive high availability so only one core pod serves at a time, with automatic failover when the leader is lost.

### Explain the Changes
This PR adds a new Kubernetes Lease-based leader election module to the core library.

The new module is responsible for acquiring and renewing leadership, releasing it on shutdown, and notifying callers when leadership is gained or lost. It provides a simple event-driven interface that can later be used by core_init and other HA-related flows.

This PR only adds the reusable leader election module and its unit tests.
The actual wiring into core_init is done separately in a follow-up PR.

New pieces:
1. Added a new Lease client in `core_lease_utils.js`. this client is provides functions to acquire, renew, and release the lease
2. added config variables CORE_LEASE_ACQUIRE_RETRY_MS and CORE_LEASE_RENEW_DEADLINE_MS to configure lease timings
3. Unit tests for the lease client 

### Issues:
1. part of https://redhat.atlassian.net/browse/RHSTOR-7491

operator side change: https://github.com/noobaa/noobaa-operator/pull/1964

### Testing Instructions:
1. `node --allow-natives-syntax ./node_modules/.bin/_mocha \
  src/test/unit_tests/util_functions_tests/test_core_lease.js`

- [ ] Doc added/updated
- [x] Tests added

## Summary

- **New Features**
  - Added Kubernetes Lease–based active/passive leader election with `leadership_acquired` and `leadership_lost` events.
  - Introduced core lease timing and Kubernetes API request timeout configuration for the leader-election behavior.
  - Expanded HTTPS trust roots by optionally loading the Kubernetes CA certificate via a new `KUBE_CA_CERTS` setting.

- **Tests**
  - Added a comprehensive leader-election unit test suite covering lease parsing/formatting, expiry/takeability logic, and acquire/renew/release edge cases.
  - Wired the new leader-election tests into the existing unit test runner.

### CODE REVIEW

File: src/util/core_lease_utils.js [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+                        return;
+                    }
+                    continue;
+                } else {
+                    dbg.warn('lease PUT failed, retrying', this._lease_name, status_code);
+                }
+                if (this._stop_renew) {
+                    break;
+                }
+                await this._interruptible_sleep(config.CORE_LEASE_ACQUIRE_RETRY_MS);
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/util/core_lease_utils.js]: this is intentional. this is a rare thing that should not even happen. in fact sleep will only increase the chance of another collision. should retry immediately to reduce release time

File: src/util/leader_election.js [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+            headers: {
+                'Content-Type': 'application/json',
+                Accept: 'application/json',
+                Authorization: `Bearer ${this._sa_token}`,
+            },
+        }, body && JSON.stringify(body), 'utf8');
+        const buffer = await read_stream_join(response);
+        const res_body = buffer.length ? JSON.parse(buffer.toString('utf8')) : {};
+        return { status_code: response.statusCode, body: res_body };
+    }
```
Comment (nadavMiz) [inline src/util/leader_election.js]: this is an existing issue. exists in kubestore as well. there shouldn't be much risk as communication is internal only. out of scope for now. should be addressed in a seperate PR

File: src/util/leader_election.js [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+            hostname: this._service_host,
+            port: this._service_port,
+            path: this._lease_path(),
+            rejectUnauthorized: false,
+            headers: {
+                'Content-Type': 'application/json',
+                Accept: 'application/json',
+                Authorization: `Bearer ${this._sa_token}`,
+            },
+        }, body && JSON.stringify(body), 'utf8');
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/util/leader_election.js]: currently make_https_request does not support timeout and no timeout is used throughout the code.  I agree this might be a problem but its out of scope for this PR. no new issue was introduced

File: src/util/core_lease_utils.js [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+     */
+    async acquire_lease() {
+        dbg.log0('acquiring core lease', this._lease_name, 'holder', this._holder);
+        for (;;) {
+            const { lease, can_take } = await this.read_lease_state();
+            if (!lease) {
+                throw new Error(`lease ${this._lease_name} not found in namespace ${this._namespace}`);
+            }
+            if (can_take) {
+                const status_code = await this.update_lease(lease, true);
```
Comment (dannyzaken) [inline src/util/core_lease_utils.js]: handle errors. e.g.: timeout, connection refused, etc. 
right now it will just throw and eventually exit the container

File: src/util/leader_election.js [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    }
+
+    /**
+     * @param {string} method
+     * @param {object} [body]
+     * @returns {Promise<{ status_code: number, body: object }>}
+     */
+    async _request(method, body) {
+        await this._init();
+        const response = await make_https_request({
```
Comment (dannyzaken) [inline src/util/leader_election.js]: we should have a reasonable timeout on the request.

File: src/cmd/core_init.js [hunk 6]
```diff
@@ -44,29 +59,174 @@ async function run_server_upgrade() {
...
+
+async function start() {
+    const version = read_package_version();
+    console.log(`Version is: ${version}`);
+    console.log('running core server init');
+
+    if (is_core_lease_enabled()) {
+        _lease_client = create_client_from_env();
+        await _lease_client.acquire_lease();
+        start_core_lease_renewal(_lease_client);
```
Comment (dannyzaken) [inline src/cmd/core_init.js]: as we discusses, maybe the flow will be clearer if you make the lease_client an event emitter. after acquire_lease add a listener for `lost_leadership` event. once got the event go into TERM sequence.

File: src/util/leader_election.js [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                    }
+                } else {
+                    dbg.warn('lease PUT failed, retrying', this._lease_name, status_code);
+                }
+                if (this._stop_renew) {
+                    break;
+                }
+                await this._interruptible_sleep(config.CORE_LEASE_ACQUIRE_RETRY_MS);
+            }
+        } finally {
```
Comment (dannyzaken) [inline src/util/leader_election.js:522]: add catch block for better error handling. Also retry with a shorter sleep if failed to access the lease.

File: config.js [hunk 8]
```diff
@@ -671,6 +671,14 @@ config.KUBE_API_ENDPOINTS = {
     TOKEN_REVIEW: '/apis/authentication.k8s.io/v1/tokenreviews'
 };
 
+// Core active-passive HA (Kubernetes Lease). Disabled when NOOBAA_CORE_LEASE_NAME is unset.
+// Time to wait between lease acquisition retries.
+config.CORE_LEASE_ACQUIRE_RETRY_MS = 3000;
+// Max time without a successful renew PUT before the leader steps down. 
+// should be less than lease duration to avoid split brain (equivalent to client-go RenewDeadline).
+//lease duration defined in the Lease yaml on the noobaa operator.
+config.CORE_LEASE_RENEW_DEADLINE_MS = 18000;
```
Comment (dannyzaken) [inline config.js]: deadline should be shorter. 18 seconds + 3 seconds for CORE_LEASE_ACQUIRE_RETRY_MS can bring us over the 20 seconds lease timeout. 
if renew errors are retried with a shorter delay (1 sec maybe), we can make the deadline much sorter.

File: src/util/leader_election.js [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+    }
+
+    async _init() {
+        if (this._initialized) return;
+        const token_buf = await fs.promises.readFile(config.KUBE_SA_TOKEN_FILE);
+        this._sa_token = token_buf.toString('utf8').trim();
+        const ns_buf = await fs.promises.readFile(config.KUBE_NAMESPACE_FILE);
+        this._namespace = ns_buf.toString('utf8').trim();
+        this._initialized = true;
+    }
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/util/leader_election.js:176]: the should not be more then one leader_election instance per pod. also it will run once sequentialy from core_init

File: src/util/leader_election.js [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+
+    /**
+     * Returns true when too long has passed since the last successful renew.
+     * Step down in this case to avoid split brain if we can no longer refresh the lease.
+     * @param {number} last_successful_renew_ms
+     * @returns {boolean}
+     */
+    _is_renew_deadline_exceeded(last_successful_renew_ms) {
+        return Date.now() - last_successful_renew_ms > config.CORE_LEASE_RENEW_DEADLINE_MS;
+    }
```
Comment (dannyzaken) [inline src/util/leader_election.js]: there is no margin for losing leadership - `CORE_LEASE_RENEW_DEADLINE_MS` (14) + `CORE_LEASE_REQUEST_TIMEOUT_MS` (5) + 'CORE_LEASE_RENEW_ERROR_SLEEP_MS' (1) = 20 seconds. We should either reduce `CORE_LEASE_RENEW_DEADLINE_MS` (maybe to 10), or account for the possible `CORE_LEASE_REQUEST_TIMEOUT_MS` when checking the deadline

File: src/util/leader_election.js [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+    /**
+     * GET the lease object from the API.
+     * @returns {Promise<object|null>}
+     */
+    async read_lease() {
+        const { status_code, body } = await this._request('GET');
+        if (status_code === 200) return body;
+        if (status_code === 404) return null;
+        throw new Error(`read lease failed status=${status_code} body=${JSON.stringify(body)}`);
+    }
```
Comment (dannyzaken) [inline src/util/leader_election.js]: All our timing constants assume that `leaseDurationSeconds` is 20 seconds. I think we should verify this  and either log or throw an error. 
Alternatively, we can change the deadline based on the lease duration, so instead of having `CORE_LEASE_RENEW_DEADLINE_MS`, we can define `LEASE_RENEW_DEADLINE_MARGIN` to be 4-5 seconds, and then `deadline = leaseDurationSeconds - CORE_LEASE_RENEW_ERROR_SLEEP_MS - CORE_LEASE_REQUEST_TIMEOUT_MS - LEASE_RENEW_DEADLINE_MARGIN`.

File: src/util/leader_election.js [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+}
+
+/**
+ * Returns true if the lease record has no valid holder (expired or empty renew time).
+ * @param {{ spec?: { holderIdentity?: string, renewTime?: string, leaseDurationSeconds?: number } }} lease
+ * @returns {boolean}
+ */
+function is_lease_expired(lease) {
+    const spec = lease?.spec || {};
+    const renew_ms = parse_lease_time(spec.renewTime);
```
Comment (dannyzaken) [inline src/util/leader_election.js]: We should consider the possibility of clock skew between the pods. It should be rare in production clusters with NTP, but in some environments it is possible. We can deal with it by not checking the expiration based on `spec.renewTime` (which is set by the other pod), but instead track the time since the lease changed its state (e.g., renewed by the holder).

File: src/util/leader_election.js [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+
+/**
+ * @returns {LeaderElector}
+ */
+function create_elector_from_env() {
+    const lease_name = process.env.NOOBAA_CORE_LEASE_NAME;
+    if (!lease_name) {
+        throw new Error('NOOBAA_CORE_LEASE_NAME is not set');
+    }
+    const holder = process.env.HOSTNAME;
```
Comment (dannyzaken) [inline src/util/leader_election.js]: We should have a unique holder identity for each instance to avoid split-brain when we have an orphaned container. We have had similar cases with the old DB statefulset. For example, when a pod is force-deleted, its container is detached from the pod and may continue running in the background for some time before being killed.

File: src/util/leader_election.js [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+     * @returns {Promise<{ status_code: number, body: object }>}
+     */
+    async _request(method, body) {
+        await this._init();
+        const response = await make_https_request({
+            method,
+            hostname: this._service_host,
+            port: this._service_port,
+            path: this._lease_path(),
+            rejectUnauthorized: false,
```
Comment (dannyzaken) [inline src/util/leader_election.js]: We should not introduce new code with `rejectUnauthorized: false`.
you can remove it and add `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt` to the `https_agent` in http_utils

File: src/util/leader_election.js [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+            const lease = await this.read_lease();
+            if (!lease) {
+                throw new Error(`lease ${this._lease_name} not found in namespace ${this._namespace}`);
+            }
+            if (!is_lease_held_by(lease, this._holder)) {
+                dbg.log0('core lease not held by us, skipping release',
+                    'holder', lease.spec?.holderIdentity, 'we are', this._holder);
+                return;
+            }
+            const status_code = await this._put_release_lease(lease);
```
Comment (dannyzaken) [inline src/util/leader_election.js:372]: add error handling for the release flow

File: src/util/leader_election.js [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+            loop_done = resolve;
+        });
+
+        try {
+            dbg.log0('starting core lease renew loop', this._lease_name);
+            let last_successful_renew_ms = Date.now();
+            while (!this._stop_renew) {
+                // lease is about to expire, step down
+                if (this._is_renew_deadline_exceeded(last_successful_renew_ms)) {
+                    dbg.error('renew deadline exceeded, shutting down pod', this._lease_name);
```
Comment (dannyzaken) [inline src/util/leader_election.js]: message should be `leadership lost` instead of `shutting down pod` which is the decision of the calling code

File: src/util/leader_election.js (hunk 11 above)
Comment (nadavMiz) [reply src/util/leader_election.js]: chose the second method as it seemed clearer and safer in case one of the times changes

File: src/util/leader_election.js (hunk 12 above)
Comment (nadavMiz) [reply src/util/leader_election.js]: right, changed it to use the observed time instead. this is also aligned with the go module. we can have a delay of a few seconds between observing the change and when it actually happened but its fine since its only used for acquire_lease, or if someone takes our lease. so we are more conservative in this way. and for our own time we still use last_successful_renew_ms, which is set immediately after we renew the lease, so we shouldn't need to change _get_renew_deadline_ms

File: src/util/leader_election.js (hunk 13 above)
Comment (nadavMiz) [reply src/util/leader_election.js]: right, this was intentional. the idea was that if the pod crash, it would not need to wait for the lease to expire before taking it. but I didn't think about the orphaned container case. changed it to have a uuid in the name

File: src/util/leader_election.js (hunk 15 above)
Comment (nadavMiz) [reply src/util/leader_election.js:372]: done, also removed the 409 handler as its not possible in our case because renew loop has already stopped at this point and it just complicates the function. worst case we will wait for the lease to expire

File: src/util/leader_election.js [hunk 17]
```diff
@@ -0,0 +1,464 @@
...
+            }
+            // Renew loop or another pod may have updated the lease between read and PUT,
+            // re-read and retry.
+            if (status_code === 409) {
+                dbg.log1('release lease 409 conflict, retrying', this._lease_name);
+                continue;
+            }
+            throw new Error(`release lease failed status=${status_code}`);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/util/leader_election.js:383]: added try catch. is this resolved to your satisfaction?

File: src/util/leader_election.js [hunk 18]
```diff
@@ -0,0 +1,495 @@
...
+            headers: {
+                'Content-Type': 'application/json',
+                Accept: 'application/json',
+                Authorization: `Bearer ${this._sa_token}`,
+            },
+        }, body && JSON.stringify(body), 'utf8');
+        const buffer = await read_stream_join(response);
+        const res_body = buffer.length ? JSON.parse(buffer.toString('utf8')) : {};
+        return { status_code: response.statusCode, body: res_body };
+    }
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/util/leader_election.js]: should be prevented by new abort controller. is this acceptable?

File: src/util/leader_election.js [hunk 19]
```diff
@@ -0,0 +1,535 @@
...
+                        // another pod acquired the lease or lease was deleted, step down
+                        dbg.error('lost core lease to another holder, leadership lost', this._lease_name, 'holder', lease?.spec?.holderIdentity, 'we are', this._holder);
+                        this.emit(EVENTS.LEADERSHIP_LOST);
+                        return;
+                    }
+                    if (this._cycle_abort.signal.aborted) continue;
+                    const status_code = await this.update_lease(lease, false, this._cycle_abort.signal);
+                    if (status_code === 200) {
+                        this._reset_renew_deadline();
+                        sleep_ms = config.CORE_LEASE_ACQUIRE_RETRY_MS;
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/util/leader_election.js:508]: at this point we are holding the lease. we always set this to the same value. this is redundant

File: src/util/leader_election.js [hunk 20]
```diff
@@ -0,0 +1,528 @@
...
+        if (!KUBERNETES_SERVICE_HOST || !KUBERNETES_SERVICE_PORT) {
+            throw new Error('LeaderElector requires in-cluster Kubernetes environment variables');
+        }
+        this._lease_name = lease_name;
+        this._holder = holder;
+        this._service_host = KUBERNETES_SERVICE_HOST;
+        this._service_port = KUBERNETES_SERVICE_PORT;
+        this._initialized = false;
+        this._stop_requested = false;
+        this._lease_duration_ms = config.CORE_LEASE_DURATION_MS;
```
Comment (dannyzaken) [inline src/util/leader_election.js]: A small inconsistency: `this._lease_duration_ms` is only used for _get_renew_deadline_ms, but for the `leaseDurationSeconds` config.CORE_LEASE_DURATION_MS is used directly.

File: src/util/leader_election.js [hunk 21]
```diff
@@ -0,0 +1,528 @@
...
+    /**
+     * Computes the renew deadline from the configured lease duration and timing constants.
+     * deadline = CORE_LEASE_DURATION_MS - CORE_LEASE_RENEW_ERROR_SLEEP_MS - CORE_LEASE_REQUEST_TIMEOUT_MS - LEASE_RENEW_DEADLINE_MARGIN_MS
+     * @returns {number} deadline in ms
+     */
+    _get_renew_deadline_ms() {
+        return this._lease_duration_ms -
+            config.CORE_LEASE_RENEW_ERROR_SLEEP_MS -
+            config.CORE_LEASE_REQUEST_TIMEOUT_MS -
+            config.LEASE_RENEW_DEADLINE_MARGIN_MS;
```
Comment (dannyzaken) [inline src/util/leader_election.js]: Worth adding a validation that this is > 0

File: src/util/leader_election.js [hunk 22]
```diff
@@ -0,0 +1,528 @@
...
+
+/**
+ * @returns {LeaderElector}
+ */
+function create_elector_from_env() {
+    const lease_name = process.env.NOOBAA_CORE_LEASE_NAME;
+    if (!lease_name) {
+        throw new Error('NOOBAA_CORE_LEASE_NAME is not set');
+    }
+    const holder = `${process.env.HOSTNAME}_${randomUUID()}`;
```
Comment (dannyzaken) [inline src/util/leader_election.js]: Just for completeness, validate that `HOSTNAME` is defined (even though this should be injected by Kubernetes automatically)

File: src/util/leader_election.js [hunk 23]
```diff
@@ -0,0 +1,528 @@
...
+        /** @type {number} local clock ms when we last saw the lease record change (clock-skew protection) */
+        this._observed_change_ms = 0;
+        /** @type {Promise<void>|null} */
+        this._running_promise = null;
+        /** @type {NodeJS.Timeout|null} */
+        this._sleep_timer = null;
+        /** @type {(() => void)|null} */
+        this._wake_sleep = null;
+        /** @type {AbortController} single controller reused across renew cycles; clearTimeout prevents abort from firing on success */
+        this._cycle_abort = new AbortController();
```
Comment (dannyzaken) [inline src/util/leader_election.js]: Why do we need`this._cycle_abort`?

File: src/util/leader_election.js (hunk 20 above)
Comment (nadavMiz) [reply src/util/leader_election.js]: changed it to always use config.CORE_LEASE_DURATION_MS. seemed more consistent with other config variables

File: src/util/leader_election.js (hunk 22 above)
Comment (nadavMiz) [reply src/util/leader_election.js]: changed it to use "noobaa-core" if undefined. this is not needed for correctness. I think randomUUID() is enough. its so it would be easier for us to understand who hold the lease. and for debugging. so that the holder would be something like:
`noobaa-core-0_8f387f47-653e-44da-b01c-fceffc973c4c`

File: src/util/leader_election.js (hunk 23 above)
Comment (nadavMiz) [reply src/util/leader_election.js]: The reason is that each renewal cycle performs three HTTP requests. In our case, each request has a 5 second timeout, so a single iteration could take up to 1 second of sleep + 3 × 5 seconds for the HTTP requests = 16 seconds before we get a chance to check whether the lease deadline has been exceeded.

The AbortController acts as a timeout for the entire renewal cycle. If the deadline is reached while a request or the sleep is still in progress, the controller aborts it immediately instead of waiting for the current iteration to finish.

I chose AbortController because it can be passed to both HTTP requests and interruptible sleep, so they can be canceled while they are running. It also works well with the 5 second timeout for HTTP requests. whichever timeout fire first aborts the request.

Comment (nadavMiz) [reply src/util/leader_election.js]: however I see that the name is not clear. I will change it to this._renew_deadline_abort

File: src/util/leader_election.js (hunk 21 above)
Comment (nadavMiz) [reply src/util/leader_election.js]: right I feel like now that we have the abort controller we don't need this calculation anymore. its also incorrect as the CORE_LEASE_REQUEST_TIMEOUT_MS is per request. I will return it to be a config variable. I will still validate this just in case

**Comment by nadavMiz** [comment]
was replaced by leader election module in the operator https://github.com/noobaa/noobaa-operator/pull/2036
can reopen if needed
